plugins/modem-manager/fu-mm-device.c

```diff
@@ -1279,10 +1217,106 @@ fu_mm_device_write_firmware_mbim_qdu(FuDevice *device,
 #endif /* MM_CHECK_VERSION(1,17,1) && MBIM_CHECK_VERSION(1,25,3) */
 
 #if MM_CHECK_VERSION(1, 17, 2)
+static gboolean
+fu_mm_device_find_edl_port(FuDevice *device, const gchar *subsystem, GError **error)
+{
+	FuMmDevice *self = FU_MM_DEVICE(device);
+
+	g_clear_pointer(&self->port_edl, g_free);
+
+	for (guint i = 0; i < 30; i++) {
+		if (fu_mm_utils_find_device_file(fu_device_get_physical_id(FU_DEVICE(self)),
+						 subsystem,
+						 &self->port_edl,
+						 NULL))
+			return TRUE;
+
+		g_usleep(250 * 1000);
+	}
+
+	g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_FAILED, "Couldn't find EDL port");
+	return FALSE;
+}
+
+static gboolean
+fu_mm_device_qcdm_switch_to_edl(FuDevice *device, gpointer userdata, GError **error)
+{
+	FuMmDevice *self = FU_MM_DEVICE(device);
+	g_autoptr(GError) error_local = NULL;
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	static const guint8 emergency_download[] = {0x4b, 0x65, 0x01, 0x00, 0x54, 0x0f, 0x7e};
+
+	locker = fu_device_locker_new_full(self,
+					   (FuDeviceLockerFunc)fu_mm_device_io_open_qcdm,
+					   (FuDeviceLockerFunc)fu_mm_device_io_close,
+					   error);
+	if (locker == NULL)
+		return fu_mm_device_find_edl_port(device, "wwan", error);
+
+	if (!fu_mm_device_qcdm_cmd(self,
+				   emergency_download,
+				   G_N_ELEMENTS(emergency_download),
+				   &error_local))
+		g_debug("couldn't send QCDM command to switch to EDL mode: %s",
+			error_local->message);
+
+	g_set_error(error,
+		    FWUPD_ERROR,
+		    FWUPD_ERROR_NOT_FOUND,
+		    "Device haven't switched to EDL yet");
+	return FALSE;
+}
+
+#if MBIM_CHECK_VERSION(1, 27, 5)
+static void
+fu_mm_device_switch_to_edl_mbim_ready(MbimDevice *device, GAsyncResult *res, GMainLoop *loop)
+{
+	g_autoptr(GError) error = NULL;
+	g_autoptr(MbimMessage) response = NULL;
+
+	response = mbim_device_command_finish(device, res, &error);
+	if (response != NULL &&
+	    mbim_message_response_get_result(response, MBIM_MESSAGE_TYPE_COMMAND_DONE, &error))
```

Comment:
this chaining makes it confusing; can we do two `if` blocks please.

Reply:
i just realized we don't to process the response so i deleted this part